### PR TITLE
Implement Simple Linter

### PR DESCRIPTION
**Pull request by dhonx**
This PR implementing simple linter, basically it just parsing stderr from `v -o /tmp/vscode_vlang/hash_name.c .`.

> Note: Warnings will be skipped & this will not linting document/file that not exists on the current workspace. 

It's Run When:
  * save document
  * active document changes

### Screenshoot
Screenshot from 2020-01-10 20-11-09
Screenshot from 2020-01-10 19-38-22
Screenshot from 2020-01-10 19-31-46

### CODE REVIEW

File: src/client/linter.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	const cwd = getCwd();
+	// const filename = relative(cwd, document.uri.fsPath);
+
+	const cmd = `-o ${outDir}lint.c .`;
+
+	let status = true;
+	let fileuri = document.uri;
+
+	execV(cmd, (err, stdout, stderr) => {
+		if (err) {
```
Comment (alexesprit) [inline src/client/linter.ts]: On Windows 10 `stderr` contains nothing. Should be something like:
```ts
const output = stderr || stdout;
...
```

File: src/client/linter.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+			collection.delete(fileuri);
+			const isWarning = stderr.substring(0, 7) === "warning";
+			if (isWarning) stderr = removeWarning(stderr);
+			const { file, line, column, message } = parseError(stderr);
+			fileuri = Uri.parse(resolve(cwd, file));
```
Comment (alexesprit) [inline src/client/linter.ts]: Should be `Uri.file` instead.

**Comment by alexesprit** [comment]
After fixing the `stderr` issue on Windows, I got the following result:
image

What I expected:
image

```
next token = `=`
v2.v:11:7: syntax error: unexpected `cdefs`, expecting `(`
    9| 
   10| const
   11| cdefs = '
             ^
   12| #define true 1
   13| #define false 0
```

File: src/client/linter.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+export function lint(document: TextDocument): boolean {
+	// Don't lint files that are not in the workspace
+	if (!workspace.getWorkspaceFolder(document.uri)) {
+		return false;
+	}
+
+	const cwd = getCwd();
+	// const filename = relative(cwd, document.uri.fsPath);
+
+	const cmd = `-o ${outDir}lint.c .`;
```
Comment (alexesprit) [inline src/client/linter.ts]: This also produces invalid errors, e.g.:
image

I changed `.` to path to file, and now actual errors are shown (see my screenshot in comments).

Comment (dhonx) [reply src/client/linter.ts]: the problem when `v -o /tmp/vscode_vlang/linter.c currentActiveFile.v` is the error is invalid
for example:
a.v
```v
struct Foo {}
```
b.v
```v 
struct Bar {}
```
sometimes the error said: `Foo` is not exist.

Comment (alexesprit) [reply src/client/linter.ts]: Aah, nevermind, it's okay.

Comment (dhonx) [reply src/client/linter.ts]: But this is still a bug  😁

Comment (alexesprit) [reply src/client/linter.ts]: This error was from an invalid file that was not supposed to be in the workspace. Aside wrong parsing, it's not an issue.

Comment (dhonx) [reply src/client/linter.ts]: Yeah, I forgot that `path` info

**Comment by alexesprit** [comment]
This feature should be toggleable to allow users to turn it on/off.

**Comment by Delta456** [comment]
Agreed

**Comment by dhonx** [comment]
Agreed !!!

**Pull request closed without merging**